React guide - first version (#32)
MarsBased/handbook: guides/react-guidelines.md

```

If using a repo for both api and client, put the above inside `client/` folder

### 2.2. <a name='Referencesprojectstructure'></a>References (project structure)

- Route definitions idea taken from [Redwood](https://github.com/redwoodjs/redwood) framework

## 3. <a name='Describemostcommonpatternsusedtosolvecommonproblems'></a>Describe most common patterns used to solve common problems

### 3.1. <a name='Statemanagement'></a>State management

- In general, prefer hooks over any other solution
- You-don-t-need-redux. But if so, use hooks interface

#### 3.1.1. <a name='Localstatemanagement'></a>Local state management

React hooks are normally enough.

#### 3.1.2. <a name='Globalstatemanagement'></a>Global state management

Prefer [zustand](https://github.com/pmndrs/zustand) over React contexts or redux.

See zustand documentation.

### 3.2. <a name='Externalservices'></a>External services

- Create a clean interface to each service. For example: `src/api/index.ts` (functions to send http requests to a REST API) or `src/graphql/index.ts` (queries and mutations of a GraphQL endpoint)
- Create a file with the types (models): `<service-name>/types.ts`

> - Use declarative data fetching: instead of `fetch` prefer `useQuery` (available both in Apollo client and react-query)

Reviewer: ```suggestion
- Use declarative data fetching: prefer `useQuery` over `fetch` (available both in Apollo client and react-query)
```

### 3.3. <a name='GraphQL'></a>GraphQL

We currently use [apollo-client](https://www.apollographql.com/docs/react/)

Try to generate types and code as much as possible.

In general keep all graphql related code inside `graphql/` folder.

### 3.4. <a name='REST'></a>REST

- A single `src/api/index.ts` exports all possible API interactions
- Use `src/api/types.ts` to declare API entity type definitions
- Client specific functionality inside `src/api/client.ts` (like, for example)
- If Auth and API are different services, is common to have two folders (`src/auth` and `src/api`) and the API depends on authorization (JWT tokens, for example). If auth and API are in the same service, the `src/auth` folder can be omitted.

### 3.5. <a name='RoutingonlyCRA'></a>Routing (only CRA)

Custom routing is only required with create react app (next.js has it's own routing standards and patterns).

- Use react-router-dom with hooks
- Create a route definitions file `src/routes.ts` with all route paths
- It helps to mimic the actual routes in the `/pages` folder

#### 3.5.1. <a name='Pagevscomponent'></a>Page vs component

A Page is a component that:

> - It is used in Router.tsx

Reviewer: ```suggestion
- It is used in `Router.tsx`
```

  admin: {
    users: () => `/admin/users`,
  },
};
```

Usage:

```tsx
import routes from "./routes";

<Link to={routes.users()}>Users</Link>;
```

#### 3.5.3. <a name='Router.tsx'></a>Router.tsx

Use the `routes` definitions to create the routes placeholders:

```tsx
<Router>
  <Switch>
    <Route exact path={routes.posts()}>
      <PostsListPage />
    </Route>
    <Route exact path={routes.post(":id")}>
      <PostPage />
    </Route>
    <Route exact path={routes.admin.users()}>
      <AdminUsersListPage />
    </Route>
  </Switch>
</Router>
```

#### 3.5.4. <a name='Accessrouteparametersusinghooks'></a>Access route parameters using hooks

For a route like `/posts/:postId/comments/:commentId` we use the following code to access route params:

```tsx
// In next version of react-router types won't be required
const { pageId, commentId } = useParams<Record<string, string>>();
```

### 3.6. <a name='Testing'></a>Testing

Follow CRA guidelines: https://reactjs.org/docs/testing.html

- Favor end-to-end-testing over components test
- More important to cover critical paths that general coverage
- Ensure business logic are pure functions and write unit test for them when needed

> When we wrote tests, we used [cypress](https://cypress.io)

Reviewer: ```suggestion
When we wrote tests, we use [cypress](https://cypress.io)
```

I guess?

## 4. <a name='Libraries'></a>Libraries

### 4.1. <a name='Recommendedlibraries'></a>Recommended libraries

- Internationalization: [react-intl](https://www.npmjs.com/package/react-intl)
- Forms: [react-hook-form](https://react-hook-form.com/)
- Global state management: [zustand](https://github.com/pmndrs/zustand)
- React query state: [react-query](https://github.com/tannerlinsley/react-query)
- Http: [xy](https://github.com/sindresorhus/ky)
- GraphQL API: [apollo-client](https://www.apollographql.com/docs/react/)
- Routing: [react-router-dom](https://reacttraining.com/react-router/web/guides/quick-start)

### 4.2. <a name='Otherlibrariesweused'></a>Other libraries we used

- Styling

>   - [tailwindcss](https://tailwindcss.com/)

Reviewer: Depends on who you ask inside MarsBased 🙈
